### PR TITLE
[backend/frontend] Enable CSV Feed Ingester (#4569)

### PR DESCRIPTION
**Pull request by Goumies**
### Proposed changes

* Add CSV Ingestion Entity
* Single-column CSV from URL content / Direct Download CSV from URL content / CSV With # Header
* Add Creation and Edition forms
* Test submitted URL and CSV Mapper

### Related issues

* https://github.com/OpenCTI-Platform/opencti/issues/4569

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

To test functionally, your opencti-worker python must be started. Then:
- Create a CSV mapper for your type of content
- Create a CSV feed ingester by navigating to http://localhost:3000/dashboard/data/ingestion/csv, select the CSV mapper
- Start it (it can be stopped, updated or deleted as well)
- Start your worker to ingest data
- After up to 10 minutes, your newly created entities should be listed on http://localhost:3000/dashboard/observations/observables

Here is some CSV Example that are available to test (single column, with header # , direct download): 
- Single Column : https://lists.blocklist.de/lists/all.txt
- With Header # : https://malsilo.gitlab.io/feeds/dumps/ip_list.txt
- With Header # with Pipe (|) column separator : https://dataplane.org/dnsrdany.txt
- With Csv direct download : https://feeds.honeynet.asia/url/latest-url-unique.csv
- With Txt File direct download : https://cybercrime-tracker.net/ccamgate.php

### CODE REVIEW

File: opencti-platform/opencti-front/src/private/components/common/form/CsvMapperField.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  return (
+    <>
+      <Field
+        component={AutocompleteField}
+        style={fieldSpacingContainerStyle}
+        name={name}
+        multiple={false}
+        textfieldprops={{
+          variant: 'standard',
+          label: t('Csv Mappers'),
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/common/form/CsvMapperField.tsx]: Far from critical, but CSV is an acronym, so I would write it in upper case only.

```suggestion
          label: t('CSV Mappers'),
```
--> + change in lang files

File: opencti-platform/opencti-front/src/private/components/data/Root.jsx [hunk 2]
```diff
@@ -120,8 +131,7 @@ const Root = () => {
           path="/dashboard/data/processing/tasks"
           component={Tasks}
         />
-      </Switch>
-    </Suspense>
+      </Switch></Suspense>
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/Root.jsx]: probably revert this change here. Format was OK I think.

**Comment by labo-flg** [comment]
Test results:
* a lot of merge conflict to handle due to changes in the i18n 
* tested OK with Single Column : https://lists.blocklist.de/lists/all.txt
* ⚠️tested KO With Header # with Pipe (|) column separator : https://dataplane.org/dnsrdany.txt --> no ingestion even after 10 minutes. I had the 2 ingesters enabled, still some messages in the backend `[OPENCTI-MODULE] CSV ingestion execution for 194 items`.
* still infinite warnings about option-value matching when creating an ingester, when selecting a mapper in the autocomplete
* errors in backend `FUNCTIONAL_ERROR: Observable is not correctly formatted` --> is it handled properly  then ?

File: opencti-platform/opencti-graphql/src/manager/ingestionManager.ts [hunk 3]
```diff
@@ -271,6 +275,80 @@ const taxiiExecutor = async (context: AuthContext) => {
...
+  if (ingestion.authentication_type === 'basic') {
+    const auth = Buffer.from(ingestion.authentication_value, 'utf-8').toString('base64');
+    headers.Authorization = `Basic ${auth}`;
+  }
+  if (ingestion.authentication_type === 'bearer') {
+    headers.Authorization = `Bearer ${ingestion.authentication_value}`;
+  }
+  let certificates;
+  if (ingestion.authentication_type === 'certificate') {
+    certificates = { cert: ingestion.authentication_value.split(':')[0], key: ingestion.authentication_value.split(':')[1], ca: ingestion.authentication_value.split(':')[0] };
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/manager/ingestionManager.ts]:
```suggestion
    const [cert, key, ca ] = ingestion.authentication_value.split(':');
    certificates = { cert, key, ca };
```

File: opencti-platform/opencti-graphql/src/manager/ingestionManager.ts [hunk 4]
```diff
@@ -271,6 +275,80 @@ const taxiiExecutor = async (context: AuthContext) => {
...
+  }
+  const httpClientOptions: GetHttpClient = { headers, rejectUnauthorized: false, responseType: 'json', certificates };
+  const httpClient = getHttpClient(httpClientOptions);
+  const { data, headers: resultHeaders } = await httpClient.get(ingestion.uri);
+  return { data, addedLast: resultHeaders['x-csv-date-added-last'] };
+};
+const csvDataToObjects = async (data: string, ingestion: BasicStoreEntityIngestionCsv, csvMapper: BasicStoreEntityCsvMapper, context: AuthContext) => { // push bundle to queues
+  const entitiesData = data.split('\n');
+  logApp.info(`[OPENCTI-MODULE] CSV ingestion execution for ${entitiesData.length} items`);
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/manager/ingestionManager.ts]: I suggest you lower the log level to debug, this happens a lot during my tests (and I only had one ingester active!)

File: opencti-platform/opencti-graphql/src/modules/ingestion/ingestion-converter.ts [hunk 5]
```diff
@@ -36,3 +36,21 @@ export const convertIngestionTaxiiToStix = (instance: StoreEntityIngestionRss):
...
 };
+
+export const convertIngestionCsvToStix = (instance: StoreEntityIngestionCsv): StixIngestionCsv => {
+  const stixObject = buildStixObject(instance);
+  return {
+    ...stixObject,
+    name: instance.name,
+    description: instance.description,
+    uri: instance.uri,
+    csvMapper_id: instance.csvMapper_id,
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/modules/ingestion/ingestion-converter.ts]: by convention, I think we use snake case in stix objects, so `csv_mapper_id` maybe?

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      variables: SINGLE_COLUMN_CSV_MAPPER
+    });
+
+    singleColumnCsvMapperId = createSingleColumnCsvMapperQueryResult?.data?.csvMapperAdd?.id;
+  });
+
+  it('should create a CSV feeds ingester', async () => {
+    const CSV_FEED_INGESTER_TO_CREATE = {
+      input: {
+        authentication_type: 'none',
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:37]: Do you think you could cover other auth methods in the tests?

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      `,
+      variables: CSV_FEED_INGESTER_TO_CREATE
+    });
+    singleColumnCsvFeedIngesterId = createSingleColumnCsvFeedsIngesterQueryResult?.data?.ingestionCsvAdd?.id;
+    expect(singleColumnCsvFeedIngesterId).toBeDefined();
+    expect(createSingleColumnCsvFeedsIngesterQueryResult?.data?.ingestionCsvAdd?.entity_type).toBe('IngestionCsv');
+    expect(createSingleColumnCsvFeedsIngesterQueryResult?.data?.ingestionCsvAdd?.ingestion_running).toBeFalsy();
+  });
+
+  it('should start the CSV feeds ingester', async () => {
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:90]: I see we do not check if we actually ingest data correctly. Might be difficult to test, I reckon.
Probably adding it at the managers levels (we only have notification-manager tests I see).
Maybe intercept the http call to the target URL to send a mock payload instead (just a few Ips for instance).

This is probably something we want to cover somehow 🤔 WDYT ?

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts (hunk 6 above)
Comment (SarahBocognano) [reply opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:37]: Yes good idea definitely need to add some !

File: opencti-platform/opencti-graphql/src/modules/ingestion/ingestion-converter.ts (hunk 5 above)
Comment (SarahBocognano) [reply opencti-platform/opencti-graphql/src/modules/ingestion/ingestion-converter.ts]: Indeed, I think you are right. Gonna change that

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts (hunk 7 above)
Comment (Goumies) [reply opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:90]: We agree. That is something we keep in mind.
We are currently working on the optimization. Right after, we will take care of the tests.

**Comment by Goumies** [comment]
> Test results:
> 
> * a lot of merge conflict to handle due to changes in the i18n
> * tested OK with Single Column : https://lists.blocklist.de/lists/all.txt
> * ⚠️tested KO With Header # with Pipe (|) column separator : https://dataplane.org/dnsrdany.txt --> no ingestion even after 10 minutes. I had the 2 ingesters enabled, still some messages in the backend `[OPENCTI-MODULE] CSV ingestion execution for 194 items`.
> * still infinite warnings about option-value matching when creating an ingester, when selecting a mapper in the autocomplete
> * errors in backend `FUNCTIONAL_ERROR: Observable is not correctly formatted` --> is it handled properly  then ?

Thanks for the feedback, @labo-flg

Errors are only logged in the backend, on purpose. Some will be listed in the ImportCSV connector jobs tracking.
As for now, CSV mappers can't distinguish several entities in a single column. If I have IPv4 and IPv6 addresses in the same column, and add those entities' representations in the selected CSV mapper, I will end with twice as many entities.

I will update this PR with the other points you have mentioned.

**Review by labo-flg — COMMENTED** [review]
I've run a complete test suite on your examples.

> Single Column : https://lists.blocklist.de/lists/all.txt
With Header # with Pipe (|) column separator : https://dataplane.org/dnsrdany.txt
With Csv direct download : https://feeds.honeynet.asia/url/latest-url-unique.csv
With Txt File direct download : https://cybercrime-tracker.net/ccamgate.php

These are OK

> With Header # : https://malsilo.gitlab.io/feeds/dumps/ip_list.txt

This one I cannot really test. The IP adresses are formatted as `ip:port` which throws `ERROR:worker:{'name': 'FUNCTIONAL_ERROR', 'message': 'Observable is not correctly formatted'}` 

Also, The CTA during CSV feed creation could be better : we do not really know why the CREATE button is disabled. TEST button should be in main 'corail' color to draw attention: CREATE won't enable until a test is done. Once this is the case, we could turn the TEST button as secondary color and put CREATE in main color.

Now this is a minor change, could be done another time if we want to merge this long-awaiting PR.

PS: I've opened an issue while testing: #5751

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx [hunk 8]
```diff
@@ -279,6 +281,18 @@ const IngestionCsvCreation: FunctionComponent<IngestionCsvCreationProps> = ({ pa
                 isOptionEqualToValue={(option: Option, value: string) => option.value === value}
                 containerStyle={fieldSpacingContainerStyle}
               />
+              <Box>
+                <div style={{ width: '100%', marginTop: 30 }}>
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx]: I think you can merge these 2 tags using the Box `sx` prop. Something like:
`<Box sx={{ width: '100%', marginTop: 30 }}>`

We try to avoid inline style as much as possible.

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvMapperTestDialog.tsx [hunk 9]
```diff
@@ -113,7 +113,7 @@ const IngestionCsvMapperTestDialog: FunctionComponent<IngestionCsvMapperTestDial
         >
           <Button
             variant="contained"
-            color="secondary"
+            color={result?.test_mapper?.nbEntities ? 'primary' : 'secondary'}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvMapperTestDialog.tsx]: My comment was more on the creation form buttons, sorry for the misunderstanding.
the "Verify" button being the first main CTA, and then the "Create" button.

The verify button in the drawer form could be color "primary" when Create is disabled, and "secondary" otherwise.
This way user open the form, put the info, sees that Create is disabled, but "Verify" is primary color so it catches the eyes as the main CTA. Once verification is ok, and Create enabled, this is the new main CTA. Create is in primary, Verify  in Secondary.

Comment (Goumies) [reply opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvMapperTestDialog.tsx]: I understand your point.
So, can you confirm that the Test button from the Ingestion Csv Mapper Test Dialog stay unchanged ?

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts (hunk 7 above)
Comment (Goumies) [reply opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:90]: I will take care of it in a complementary PR

File: opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts (hunk 6 above)
Comment (Goumies) [reply opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/ingestion-csv-test.ts:37]: I will take care of it in a complementary PR

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvMapperTestDialog.tsx (hunk 9 above)
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvMapperTestDialog.tsx]: yes it can stay unchanged IMO. The UI on this dialog is pretty clear I think.

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+                label={t_i18n('CSV URL')}
+                fullWidth={true}
+                style={fieldSpacingContainerStyle}
+              />
+              {
+                queryRef
+                && <React.Suspense fallback={<Loader variant={LoaderVariant.inElement} />}>
+                  <CsvMapperField
+                    name="csv_mapper_id"
+                    isOptionEqualToValue={(option: Option, value: string) => option.value === value}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx]:
```suggestion
                    isOptionEqualToValue={(option: Option, { value }: Option) => option.value === value}
```
and change accordingly the prop type in CSVMapperField

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                component={TextField}
+                variant="standard"
+                name="uri"
+                label={t_i18n('CSV URL')}
+                fullWidth={true}
+                style={fieldSpacingContainerStyle}
+              />
+              {
+                queryRef
+                && <React.Suspense fallback={<Loader variant={LoaderVariant.inElement} />}>
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx:196]:
```suggestion
                queryRef && (
                  <React.Suspense fallback={<Loader variant={LoaderVariant.inElement} />}>
```

Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvCreation.tsx]: strange indentation

usually we do 

```jsx
queryRef && (
  <React.Suspense fallback={<Loader variant={LoaderVariant.inElement} />}>
  ....
  </React.Suspense>
)
```

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvEdition.tsx [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+    user_id: convertUser(ingestionCsvData, 'user'),
+    references: undefined,
+  };
+
+  const queryRef = useQueryLoading<CsvMapperFieldSearchQuery>(csvMapperQuery);
+
+  return (
+    <Formik<IngestionCsvEditionForm>
+      enableReinitialize={true}
+      initialValues={initialValues as never}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvEdition.tsx]:
```suggestion
      initialValues={initialValues}
```

File: opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvEdition.tsx [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+          {
+            queryRef
+            && <React.Suspense fallback={<Loader variant={LoaderVariant.inElement} />}>
+              <CsvMapperField
+                name="csv_mapper_id"
+                isOptionEqualToValue={(option: Option, value: string) => option.value === value}
+                onChange={handleSubmitField}
+                queryRef={queryRef}
+              />
+            </React.Suspense>
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/data/ingestionCsv/IngestionCsvEdition.tsx]: same remarks as in creation form

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql [hunk 14]
```diff
@@ -743,7 +743,7 @@ type Work {
   received_time: DateTime
   processed_time: DateTime
   completed_time: DateTime
-  completed_number: Int
+  completed_number: String
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/config/schema/opencti.graphql]: are you sure about this?
Might collide with Julien's work on #5762  ?

File: opencti-platform/opencti-graphql/src/domain/work.js [hunk 15]
```diff
@@ -189,7 +189,7 @@ export const createWork = async (context, user, connector, friendlyName, sourceI
     received_time: receivedTime,
     processed_time: null,
     completed_time: null,
-    completed_number: 0,
+    completed_number: '0',
```
Comment (labo-flg) [inline opencti-platform/opencti-graphql/src/domain/work.js]: Same remark, change of type

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql (hunk 14 above)
Comment (Goumies) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql]: Let me try it out.

Comment (Goumies) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql]: Since the previous error is no longer logged, I have rollbacked this modification

File: opencti-platform/opencti-graphql/src/domain/work.js (hunk 15 above)
Comment (Goumies) [reply opencti-platform/opencti-graphql/src/domain/work.js]: Same here

**Review by labo-flg — APPROVED** [review]
